File: src/contest/ioi/IOI_2015_Teams.cpp
#include <bits/stdc++.h>

#define mp make_pair
#define pb push_back
#define INF 1<<30
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define SIZE 500100
#define l(x) x << 1
#define r(x) x << 1|1
#define m(x, y) (x + y)/2

using namespace std;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;

struct Node {
    int left, right, cnt;
    Node () {}
    Node (int cnt) {
        this->cnt = cnt;
    }
    Node (int left, int right, int cnt) {
        this->left = left;
        this->right = right;
        this->cnt = cnt;
    }
};

int N;
pi student[SIZE];
vector<int> a, b;

int seg[SIZE], used;
Node nodes[10000000];

int nodeIndex;

int update (int prev, int l, int r, int val, int inc) {
    if (l <= val && val <= r) {
        if (l == r) {
            nodes[nodeIndex].cnt = nodes[prev].cnt + inc;
            return nodeIndex++;
        }
        int mid = (l + r) >> 1;
        int curr = nodeIndex++;
        nodes[curr].cnt = nodes[prev].cnt + inc;
        nodes[curr].left = update(nodes[prev].left, l, mid, val, inc);
        nodes[curr].right = update(nodes[prev].right, mid + 1, r, val, inc);
        return curr;
    }
    return prev;
}

int getSum (int curr, int used, int l, int r, int ql, int qr) {
    if (l == ql && r == qr)
        return nodes[curr].cnt - nodes[used].cnt;

    int mid = (l + r) >> 1;
    if (qr <= mid)
        return getSum(nodes[curr].left, nodes[used].left, l, mid, ql, qr);
    else if (ql > mid)
        return getSum(nodes[curr].right, nodes[used].right, mid + 1, r, ql, qr);
    return getSum(nodes[curr].left, nodes[used].left, l, mid, ql, mid) + getSum(nodes[curr].right, nodes[used].right, mid + 1, r, mid + 1, qr);
}

pi query (int curr, int used, int K, int l, int r, int ql, int qr) {
    int mid = (l + r) >> 1;
    if (K == 0)
        return mp(0, used);

    if (l == ql && r == qr) {
        if (l == r) {
            if (nodes[curr].cnt - nodes[used].cnt <= K) {
                return mp(nodes[curr].cnt - nodes[used].cnt, curr);
            }
            nodes[nodeIndex].cnt = K + nodes[used].cnt;
            nodes[nodeIndex].left = -1;
            nodes[nodeIndex].right = -1;
            return mp(K, nodeIndex++);
        }

        if (nodes[curr].cnt - nodes[used].cnt <= K)
            return mp(nodes[curr].cnt - nodes[used].cnt, curr);
        else {
            int leftSz = nodes[nodes[curr].left].cnt - nodes[nodes[used].left].cnt;
            if (leftSz < K) {
                pi right = query(nodes[curr].right, nodes[used].right, K - leftSz, mid + 1, r, mid + 1, qr);
                nodes[nodeIndex].left = nodes[curr].left;
                nodes[nodeIndex].right = right.second;
                nodes[nodeIndex].cnt = nodes[nodes[curr].left].cnt + nodes[right.second].cnt;
                return mp(leftSz + right.first, nodeIndex++);
            } else {
                pi left = query(nodes[curr].left, nodes[used].left, K, l, mid, ql, mid);
                nodes[nodeIndex].left = left.second;
                nodes[nodeIndex].right = nodes[used].right;
                nodes[nodeIndex].cnt = nodes[left.second].cnt + nodes[nodes[used].right].cnt;
                return mp(K, nodeIndex++);
            }
        }
    }

    if (qr <= mid) {
        pi s = query(nodes[curr].left, nodes[used].left, K, l, mid, ql, qr);
        nodes[nodeIndex].left = s.second;
        nodes[nodeIndex].right = nodes[used].right;
        nodes[nodeIndex].cnt = nodes[s.second].cnt + nodes[nodes[used].right].cnt;
        return mp(s.first, nodeIndex++);
    } else if (ql > mid) {
        pi s = query(nodes[curr].right, nodes[used].right, K, mid + 1, r, ql, qr);
        nodes[nodeIndex].left = nodes[used].left;
        nodes[nodeIndex].right = s.second;
        nodes[nodeIndex].cnt = nodes[nodes[used].left].cnt + nodes[s.second].cnt;
        return mp(s.first, nodeIndex++);
    } else {
        pi left = query(nodes[curr].left, nodes[used].left, K, l, mid, ql, mid);
        pi right = query(nodes[curr].right, nodes[used].right, K - left.first, mid + 1, r, mid + 1, qr);
        nodes[nodeIndex].left = left.second;
        nodes[nodeIndex].right = right.second;
        nodes[nodeIndex].cnt = nodes[left.second].cnt + nodes[right.second].cnt;
        return mp(left.first + right.first, nodeIndex++);
    }
}

void init (int n, int A[], int B[]) {
    N = n;
    a.pb(0);
    b.pb(0);

    for (int i = 0; i < N; i++) {
        a.pb(A[i]);
        a.pb(B[i]);
        student[i] = mp(A[i], B[i]);
    }

    sort(a.begin(), a.end());
    a.erase(unique(a.begin(), a.end()), a.end());

    sort(student, student + N);

    nodes[nodeIndex].left = nodes[nodeIndex].right = nodeIndex;
    seg[0] = nodeIndex++;

    nodes[nodeIndex].left = nodes[nodeIndex].right = nodeIndex;
    used = nodeIndex++;

    int index = 0;

    for (int i = 1; i < (int)a.size(); i++) {
        int prev = seg[i - 1];
        while (index < N && student[index].first <= a[i]) {
            int lo = upper_bound(a.begin(), a.end(), student[index].second) - 1 - a.begin();
            prev = update(prev, 0, a.size() - 1, lo, 1);
            index++;
        }
        seg[i] = prev;
    }
}

bool can (int M, int K[]) {
    int currIndex = nodeIndex;
    nodes[nodeIndex].left = nodes[nodeIndex].right = nodeIndex;
    nodes[nodeIndex].cnt = 0;
    used = nodeIndex++;

    int sz = M;
    int val[sz];

    for (int j = 0; j < sz; j++)
        val[j] = K[j];

    sort(val, val + sz);

    bool valid = true;
    for (int j = 0; j < sz; j++) {
        int curr =  upper_bound(a.begin(), a.end(), val[j]) - 1 - a.begin();
        int next = lower_bound(a.begin(), a.end(), val[j]) - a.begin();

        if (next > (int)a.size() - 1) {
            valid = false;
            break;
        }

        int sum = getSum(seg[curr], used, 0, a.size() - 1, next, a.size() - 1);

        if (sum < val[j]) {
            valid = false;
            break;
        }

        used = query(seg[curr], used, val[j], 0, a.size() - 1, next, a.size() - 1).second;
    }
    nodeIndex = currIndex;
    if (valid) {
        return 1;
    } else {
        return 0;
    }
}

int main () {
    int n, m;
    scanf("%d", &n);
    int A[n], B[n];

    for (int i = 0; i < n; i++)
        scanf("%d%d", &A[i], &B[i]);

    init(n, A, B);

    scanf("%d", &m);

    for (int i = 0; i < m; i++) {
        int cnt;
        scanf("%d", &cnt);
        int K[cnt];
        for (int i = 0; i < cnt; i++)
            scanf("%d", &K[i]);
        printf("%d\n", can(cnt, K));
    }
}

File: src/contest/ioi/IOI_2015_Teams.java
package contest.ioi;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.StringTokenizer;

public class IOI_2015_Teams {

  static BufferedReader br;
  static OutputStreamWriter out;
  static StringTokenizer st;

  static Person[] p;

  static int N, M;

  static int[] seg;
  static int used;

  static ArrayList<Integer> a;

  static Node[] nodes = new Node[12500000];
  static int nodeIndex = 0;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    out = new OutputStreamWriter(System.out);
    //br = new BufferedReader(new FileReader("in.txt"));
    //out = new PrintWriter(new FileWriter("out.txt"));

    N = readInt();

    p = new Person[N];

    a = new ArrayList<Integer>();

    a.add(0);

    for (int i = 0; i < N; i++) {
      int lo = readInt();
      int hi = readInt();
      p[i] = new Person(lo, hi);
      a.add(lo);
      a.add(hi);
    }

    Collections.sort(a);
    a = unique(a);

    seg = new int[a.size()];
    for (int i = 0; i < 12500000; i++)
      nodes[i] = new Node(0);

    Arrays.sort(p);

    nodes[nodeIndex].left = nodes[nodeIndex].right = nodeIndex;
    seg[0] = nodeIndex++;

    int index = 0;

    for (int i = 1; i < a.size(); i++) {
      int prev = seg[i - 1];
      while (index < N && p[index].lo <= a.get(i)) {
        prev = update(prev, 0, a.size() - 1, findFloor(a, p[index].hi), 1);
        index++;
      }

      seg[i] = prev;
    }

    M = readInt();

    main:
    for (int i = 0; i < M; i++) {
      int currIndex = nodeIndex;
      nodes[nodeIndex].left = nodes[nodeIndex].right = nodeIndex;
      used = nodeIndex++;

      int sz = readInt();
      int[] val = new int[sz];

      for (int j = 0; j < sz; j++)
        val[j] = readInt();

      Arrays.sort(val);
      for (int j = 0; j < sz; j++) {
        int curr = findFloor(a, val[j]);
        int next = findCeil(a, val[j]);

        if (next > a.size() - 1) {
          out.write('0');
          out.write('\n');
          nodeIndex = currIndex;
          continue main;
        }

        int sum = getSum(seg[curr], used, 0, a.size() - 1, next, a.size() - 1);

        if (sum < val[j]) {
          out.write('0');
          out.write('\n');
          nodeIndex = currIndex;
          continue main;
        }

        used = query(seg[curr], used, val[j], 0, a.size() - 1, next, a.size() - 1).n;
      }
      out.write('1');
      out.write('\n');
      nodeIndex = currIndex;
    }

    out.close();
  }

  static int getSum(int curr, int used, int l, int r, int ql, int qr) {
    if (l == ql && r == qr)
      return nodes[curr].cnt - nodes[used].cnt;

    int mid = (l + r) >> 1;
    if (qr <= mid)
      return getSum(nodes[curr].left, nodes[used].left, l, mid, ql, qr);
    else if (ql > mid)
      return getSum(nodes[curr].right, nodes[used].right, mid + 1, r, ql, qr);
    return getSum(nodes[curr].left, nodes[used].left, l, mid, ql, mid) + getSum(nodes[curr].right, nodes[used].right, mid + 1, r, mid + 1, qr);
  }

  static State query(int curr, int used, int K, int l, int r, int ql, int qr) {
    int mid = (l + r) >> 1;
    if (K == 0)
      return new State(0, used);

    if (l == ql && r == qr) {
      if (l == r) {
        if (nodes[curr].cnt - nodes[used].cnt <= K) {
          return new State(nodes[curr].cnt - nodes[used].cnt, curr);
        }
        nodes[nodeIndex].cnt = K + nodes[used].cnt;
        nodes[nodeIndex].left = -1;
        nodes[nodeIndex].right = -1;
        return new State(K, nodeIndex++);
      }

      if (nodes[curr].cnt - nodes[used].cnt <= K)
        return new State(nodes[curr].cnt - nodes[used].cnt, curr);
      else {
        int leftSz = nodes[nodes[curr].left].cnt - nodes[nodes[used].left].cnt;
        if (leftSz < K) {
          State right = query(nodes[curr].right, nodes[used].right, K - leftSz, mid + 1, r, mid + 1, qr);
          assert leftSz + right.cnt == nodes[nodes[curr].left].cnt + nodes[right.n].cnt - nodes[used].cnt;
          nodes[nodeIndex].left = nodes[curr].left;
          nodes[nodeIndex].right = right.n;
          nodes[nodeIndex].cnt = nodes[nodes[curr].left].cnt + nodes[right.n].cnt;
          return new State(leftSz + right.cnt, nodeIndex++);
        } else {
          State left = query(nodes[curr].left, nodes[used].left, K, l, mid, ql, mid);
          assert K == nodes[left.n].cnt + nodes[nodes[used].right].cnt - nodes[used].cnt;
          nodes[nodeIndex].left = left.n;
          nodes[nodeIndex].right = nodes[used].right;
          nodes[nodeIndex].cnt = nodes[left.n].cnt + nodes[nodes[used].right].cnt;
          return new State(K, nodeIndex++);
        }
      }
    }

    if (qr <= mid) {
      State s = query(nodes[curr].left, nodes[used].left, K, l, mid, ql, qr);
      nodes[nodeIndex].left = s.n;
      nodes[nodeIndex].right = nodes[used].right;
      nodes[nodeIndex].cnt = nodes[s.n].cnt + nodes[nodes[used].right].cnt;
      return new State(s.cnt, nodeIndex++);
    } else if (ql > mid) {
      State s = query(nodes[curr].right, nodes[used].right, K, mid + 1, r, ql, qr);
      nodes[nodeIndex].left = nodes[used].left;
      nodes[nodeIndex].right = s.n;
      nodes[nodeIndex].cnt = nodes[nodes[used].left].cnt + nodes[s.n].cnt;
      return new State(s.cnt, nodeIndex++);
    } else {
      State left = query(nodes[curr].left, nodes[used].left, K, l, mid, ql, mid);
      State right = query(nodes[curr].right, nodes[used].right, K - left.cnt, mid + 1, r, mid + 1, qr);
      nodes[nodeIndex].left = left.n;
      nodes[nodeIndex].right = right.n;
      nodes[nodeIndex].cnt = nodes[left.n].cnt + nodes[right.n].cnt;
      return new State(left.cnt + right.cnt, nodeIndex++);
    }
  }

  static int findFloor(ArrayList<Integer> val, int x) {
    int lo = 0;
    int hi = val.size() - 1;
    while (lo <= hi) {
      int mid = (lo + hi) >> 1;
      if (val.get(mid) <= x) {
        lo = mid + 1;
      } else {
        hi = mid - 1;
      }
    }
    return hi;
  }

  static int findCeil(ArrayList<Integer> val, int x) {
    int lo = 0;
    int hi = val.size() - 1;
    while (lo <= hi) {
      int mid = (lo + hi) >> 1;
      if (val.get(mid) < x) {
        lo = mid + 1;
      } else {
        hi = mid - 1;
      }
    }
    return lo;
  }

  static ArrayList<Integer> unique(ArrayList<Integer> val) {
    ArrayList<Integer> ret = new ArrayList<Integer>();
    if (val.isEmpty())
      return ret;
    ret.add(val.get(0));

    for (int i = 1; i < val.size(); i++)
      if (val.get(i) != val.get(i - 1))
        ret.add(val.get(i));

    return ret;
  }

  static int update(int prev, int l, int r, int val, int inc) {
    if (l <= val && val <= r) {
      if (l == r) {
        nodes[nodeIndex].cnt = nodes[prev].cnt + inc;
        return nodeIndex++;
      }
      int mid = (l + r) >> 1;
      int curr = nodeIndex++;
      nodes[curr].cnt = nodes[prev].cnt + inc;
      nodes[curr].left = update(nodes[prev].left, l, mid, val, inc);
      nodes[curr].right = update(nodes[prev].right, mid + 1, r, val, inc);
      return curr;
    }
    return prev;
  }

  static String next() throws IOException {
    while (st == null || !st.hasMoreTokens())
      st = new StringTokenizer(br.readLine().trim());
    return st.nextToken();
  }

  static long readLong() throws IOException {
    return Long.parseLong(next());
  }

  static int readInt() throws IOException {
    return Integer.parseInt(next());
  }

  static double readDouble() throws IOException {
    return Double.parseDouble(next());
  }

  static char readCharacter() throws IOException {
    return next().charAt(0);
  }

  static String readLine() throws IOException {
    return br.readLine().trim();
  }

  static class State {
    int cnt;
    int n;

    State(int cnt, int n) {
      this.cnt = cnt;
      this.n = n;
    }
  }

  static class Person implements Comparable<Person> {
    int lo, hi;

    Person(int lo, int hi) {
      this.lo = lo;
      this.hi = hi;
    }

    @Override
    public int compareTo(Person p) {
      return lo - p.lo;
    }
  }

  static class Node {
    int left, right;
    int cnt;

    Node(int cnt) {
      this(-1, -1, cnt);
    }

    Node(int left, int right, int cnt) {
      this.left = left;
      this.right = right;
      this.cnt = cnt;
    }
  }
}
